model/rules.py

```diff
@@ -0,0 +1,45 @@
+from enum import Enum
+import abstract.guards as guards
+
+
+class RuleType(Enum):
+    """Collection of rule types used to uniquely identify each subclass of Rule. The Enum's value is what should show up in the rule YAML file"""
+    COMMIT_MESSAGE_SIZE = "CommitMessageSize"
+
+
+class Rule:
+    """Pseudo-abstract base data class for all concrete rules. A rule is defined by the rule type which gets used to downcast the concrete rule type at runtime."""
+
+    def __init__(self, type: RuleType) -> None:
+        guards.prevent_abstract_instantiation(self, Rule)
+        self.type = type
+
+    def __repr__(self) -> str:
+        """This is to allow all rules to be printable in a human-readable manner for debugging
+
+        Never call this method explicitely, this is Python's way of converting an object to string, which gets automatically
```

Comment:
*explicitly
In my opinion, you can make a shorter version of this, since it will not be called explicitly, might as well just explain why it is here, "Automatically called after `print(my_rule)` (....)", might be enough, wdyt?

Reply:
Changed to:
> Enum which identifies the rule's type. The Enum's value is what should show up in the rule YAML file